Copyedit agent-based-model-communication-networks (#646)
programminghistorian/ph-submissions: en/drafts/originals/agent-based-model-communication-networks.md

plt.show()
```

{% include figure.html filename="en-or-agent-based-model-communication-networks-03.png" alt="A histogram of the agents and the number of letters they received. The y-axis displays the number of agents who have received a certain number of letters, displayed on the x-axis." caption="Figure 3. Histogram of the letters received by all agents after 100 model runs" %}

This runs 100 instantiations of the model, each of which runs for 10 steps. (Notice that we set the histogram [bins](https://en.wikipedia.org/wiki/Data_binning) to integers, since agents can only receive whole numbers of letters.) By running the model 100 times, we smooth out some of the 'noise' of randomness, and approach the model’s overall expected behavior. For now, the letter distribution looks close to a [normal distribution (or bell curve)](https://en.wikipedia.org/wiki/Normal_distribution), which is expected since the process is random.

> _Bonus question 2_:

> Can you rewrite the function above to plot the number of sent letters? What histogram do you expect for that? *Hint*: Are agents sending letters in every round?

Author: Could you please clarify what you mean by 'the function above'?

Reviewer: We meant the previous code block where the plot is coded - maybe function is misleading here, then. I'd suggest to simply replace it then, so something like:  "Can you rewrite the code block above ..."

Reviewer: ```suggestion
Can you rewrite the code block above to plot the number of sent letters? What histogram do you expect for that? *Hint*: Are agents sending letters in every round?
```

#### 2.5.5 Adding Space

Let's add some more realistic behavior by introducing space between the agents, which should influence their letter-sending decisions. `Mesa` currently supports two kinds of spaces: `grid` and `continuous`. Grids are divided into cells, and agents exist on one particular cell, like pieces on a chess board. Continuous space, in contrast, allows agents to take any arbitrary position. Both grids and continuous spaces are frequently [toroidal](https://en.wikipedia.org/wiki/Toroid), meaning that the edges of this 'world' wrap around, with cells on the right edge connected to those on the left edge, and the top to the bottom. This prevents some cells having fewer neighbors than others, or agents being able to fall off the edge of the environment.

Let’s add a simple spatial element to our model by putting our agents on a grid, and making them move around it at random. Instead of sending a letter to any random agent, they’ll give it to an agent on the same cell. We could imagine that this represents them existing close enough to know each other and have reasons to send a letter in the first place.

> `Mesa` offers two main types of grids: `SingleGrid` and `MultiGrid`.[^22] `SingleGrid` enforces at most one agent per cell; `MultiGrid` allows multiple agents in the same cell. Since we want agents to be able to share a cell, we use `MultiGrid`, which we instatiate with width and height paramaters, and which will always be toroidal:

Author: I was a little confused by the phrasing 'which we instantiate with width and height parameters': do we set a certain width and height here, or do we just create the possibility of setting a width and height?

Also, am I correct that 'which will always be toroidal' refers to the parameter set to `True`? If so, perhaps we can clarify that too.

Reviewer: It's about the possibility of width and height, so as arguments of the model function, see §98 

Regarding toroida spacel: exactly! I'd suggest "...and which will always be toroidal (hence the 'true' designation)"

Reviewer: ```suggestion
`Mesa` offers two main types of grids: `SingleGrid` and `MultiGrid`.[^22] `SingleGrid` enforces at most one agent per cell; `MultiGrid` allows multiple agents in the same cell. Since we want agents to be able to share a cell, we use `MultiGrid`, which we instatiate with width and height paramaters, and which will always be toroidal (hence the 'true' designation):
```

        self.model.grid.move_agent(self, new_position)
```

> Next, we need to find all the other agents present in our agent's new cell, and send one of them a letter. We can get the contents of one or more cells using the grid's `get_cell_list_contents` method, or by accessing a cell directly. The method accepts either a single coordinate tuple, or a list of tuples.

Author: What do you mean by 'accessing a cell directly'?

Reviewer: @maltevogl what method did you mean exactly? I'm not sure

Reviewer 2: Cells can also be accessed by an index. So we could extend the text by : "or by accessing a cell directly by its internal number."

Author: Sorry @maltevogl, I'm still a little lost. When you say "accessing a cell", do you mean retrieving a list of the cell's contents? By "internal number", do you mean the cell coordinates? 

In the sentence that comes just after, is "The method" the first method (`get_cell_list_contents`) or the second method (accessing a cell)?

Reviewer 2: No problem at all. I think it is not really necessary to keep this part of the sentence if it leads to confusion. The get_cell_list_contents method is meant to be used. So if you are fine with it, I would suggest to remove the part after the comma: ", or by accessing a cell directly." Would that be OK? That would also solve the problem with the next sentence, since there is then only one method.

Author: Thanks @maltevogl, I think that's a good solution. It's clearer just to focus on one method!

```

> {% include figure.html filename="en-or-agent-based-model-communication-networks-04.png" alt="Two-dimensional grid where each cell is colored based on how many agents are present on it, on a scale of 0 to 3, with lighter yellow colors indicating more agents and darker blue colors indicating fewer agents." caption="Figure 4. Color mesh showing the number of agents present on each cell of our grid space" %}

Author: I noticed that there is a small typo in the figure's title 'Number of Agents presents in Cell'.

Could you please send me a new figure with the corrected title, perhaps something like 'Number of Agents Present in Each Cell'? You can email it to me directly, publishing.assistant[@]programminghistorian.org

Reviewer: @maltevogl

Reviewer 2: I ll generate a new figure and send it to you.

Author: Thank you!

> _Bonus question 3_:

> Letters are sent to direct neighbors. How could you implement sending letters only to agents far apart, e.g. with at least a distance of three cells? *Hint*: You will have to define a distance measure on grids. See for example this [_Programming Historian_ lesson on similarity measures](https://programminghistorian.org/en/lessons/common-similarity-measures#city-block-manhattan-distance).

Author: 'Letters are sent to direct neighbours': I think you mean that an agent's direct neighbour is an agent who shares the same cell, but this is confusing because we've recently been speaking of cell neighbours as the cells right next to a given cell. 

Is there a different way we could phrase this question?

Reviewer: I'd suggest "Letters are sent to agents **in neighboring cells**. How could you implement sending letters only to agents **further apart than that**, e.g., ...", that should be fine, wouldn't it?

Reviewer: ```suggestion
Letters are sent to agents in neighboring cells. How could you implement sending letters only to agents further apart than that, e.g. with at least a distance of three cells? *Hint*: You will have to define a distance measure on grids. See for example this [_Programming Historian_ lesson on similarity measures](https://programminghistorian.org/en/lessons/common-similarity-measures#city-block-manhattan-distance).
```

Author: Hmm, I think I may not be understanding the agents' step method correctly then. I was under the impression that up until now, agents would 1. move to a neighbouring cell, then 2. send a letter to an agent that shared that cell.

(From these sentences:
"Now we need to add to the agents’ behaviors, letting them move around and only send letters to other agents in the same cell. [...] First, let’s handle movement, and have the agents move to a neighboring cell. [...] Next, we need to get all the other agents present in a cell, and send one of them a letter.")

Or perhaps by "letters are sent to agents in neighbouring cells" (in the Bonus Question here) you mean that those agents were in the _previously_ neighbouring cell, i.e. before the sender moved to that cell? If that's the case, I can work around it to make that clearer.

Apologies – I just want to make sure I'm not introducing any errors to the lesson!

Reviewer: Oof, of course you're right, I'm sorry. That's what I get from suggesting things while in a meeting... Does this make more sense?

```suggestion
Letters are sent to direct neighbors inside the same cell. How could you implement sending letters only to agents that are further away, e.g. with at least a distance of three cells? *Hint*: You will have to define a distance measure on grids. See for example this [_Programming Historian_ lesson on similarity measures](https://programminghistorian.org/en/lessons/common-similarity-measures#city-block-manhattan-distance).
```

Author: Thank you, that's very helpful!

```

After every step of the model, the data collector will collect and store each agent's `letters_sent` and `letters_received` values.

We run the model just as we did above:

```python
model = LetterModel(50, 10, 10)
for i in range(100):
    model.step()
```

> The `DataCollector` can export the data it has collected as a pandas `DataFrame`, allowing for easy interactive analysis. We can get the `agent_letters` data like this:

Author: Could we add a few words to explain/confirm what the `agent_letters` data refers to?

Reviewer: ```suggestion
The `DataCollector` can export the data it has collected as a pandas `DataFrame`, allowing for easy interactive analysis. We can get the `agent_letters` data, which currently records information on letters received and letters sent as well as the total number of letters received by all agents, like this:
```

Reviewer: would this suffice? or would you prefer a separate sentence for readability?

Author: I'm still slightly unsure. Does the `agent_letters` data contain the number of letters sent and received by each agent at each step, as well as the combination of these numbers i.e. the total number of letters sent and the total number of letters received by each agent at the end of the simulation?

This total seems to be what Figure 5 shows below: the "Number of letters sent and received by a selection of agents, at step 100 of the simulation".

Reviewer: Yes, almost correct! On the agent level (ie, per agent) letters sent and received are collected; on the model level (ie, for everything) the total number of "letters received" is collected. For the latter, you can look for the compute_received_letters function.

```python
self.datacollector = mesa.DataCollector(
            agent_reporters={ -> per agent
                "Letters_sent": "letters_sent", 
                "Letters_received": "letters_received"
            },
            model_reporters={"All letters":compute_received_letters} -> for the whole model / every agent combined
        )
```
---

```python
def compute_received_letters(model): -> the total / model based amount
    number_of_received_letters = 0
    for agent in model.schedule.agents:
        number_of_received_letters += agent.letters_received
    return number_of_received_letters
```

Author: Thanks @merrygin – I've changed it to:

We can get the `agent_letters` data (the combined agent-level and model-level data) like this: [...]

Does that make sense?

Reviewer: Yes that's great

plt.show()
```

{% include figure.html filename="en-or-agent-based-model-communication-networks-06.png" alt="Histogram where the y-axis shows the number of agents, and the x-axis shows how many letters were sent by those agents. The numbers of letters sent range from about 25 to over 50." caption="Figure 6. Histogram of letters sent by agents after 100 steps of simulating the model." %}

You can also use `pandas` to export the data to a CSV ([comma-separated value](https://en.wikipedia.org/wiki/Comma-separated_values)) file, which can be opened by any common spreadsheet application.

If you do not specify a file path, the file will be saved in the local directory. After you run the code below, you will see the file `agent_data.csv` appear.

```python
agent_letters.to_csv("agent_data.csv")
```

> Having exported the data, we can now follow several approaches to test the hypotheses that we initially encoded in the model. The goal to systematically test - or validate - a model is to check if the model actually represents what it is supposed to. This can range from simply comparing your expectations to the outputs; to analyzing the internal consistency of the model over a detailed exploration of the possible parameters of your simulation (a so-called 'parameter space'); all the way to a detailed calibration to available empirical data. Mehdizadeh et al 2022, p. 8-9, writing from the perspective of mobility studies, offer a sensible differentiation of various validation methods, as well as good examples of how to evaluate Agent-Based Models in other fields.[^23]

Author: I struggled a little with the sentence that starts 'This can range from [...].' Especially, I wasn't sure what you meant by 'analyzing the internal consistency of the model over a detailed exploration of the possible parameters of your simulation', or 'a detailed calibration to available empirical data'.

I imagine these are complex methods which are difficult to describe in just a few words, but would you be able to try rephrasing them slightly, to make them a little clearer? It might help to break that long sentence up.

Reviewer: ```suggestion
Having exported the data, we can now follow several approaches to test the hypotheses that we initially encoded in the model. The goal to systematically test - or validate - a model is to check if the model actually represents what it is supposed to. This can range from simply comparing your expectations to the outputs to systematically analyzing the internal consistency of the model. Calibrating your model to available empirical data is one of the most simple and robust methods of testing, but of course this might be challenging due to data availability. A more involved method is to explore the effects of a wide range of possible parameters of your simulation in detail (a so-called 'parameter space').  Mehdizadeh et al 2022, p. 8-9, writing from the perspective of mobility studies, offer a sensible differentiation of various validation methods, as well as good examples of how to evaluate Agent-Based Models in other fields.[^23]
```

Reviewer: I restructured and extended this section a bit and hope this is a bit more clear now - you're totally right that this was hard to read before.

Author: Thank you very much!

    )[0]
```

We'll also reduce the rate of agents moving to different cell at every step, as people don't constantly relocate. For this, we introduce another weighted random choice, this time with fixed weight. Now, the chance of an agent moving will only be 20%, or whenever they draw a '1'.

```python
if self.random.choices([0,1], weights=[0.8, 0.2], k=1)[0] == 1:
    new_position = self.random.choice(possible_steps)
    self.model.grid.move_agent(self, new_position)
```

> To initialize the agents with this new option, we have to add a final parameter to the model itself. All together, we get the following new definitions for agents and model:

Author: You had written 'we also have to add another parameter' – what is this parameter?

Reviewer: @maltevogl

Reviewer 2: The parameter is reinforce = False to switch between two modes of running the model. I suggest to add at the end of the sentence:
 
"To initialize the agents with this new option, we have to add a final parameter to the model itself, which controls if the reinforcement process should be used for selecting letter receivers." 

Sorry I could not figure out how to suggest elegantly...
